src/librustc_macros/src/query.rs

```diff
@@ -0,0 +1,320 @@
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use syn::{
+    Token, Ident, Type, Attribute, ReturnType, Expr, Block,
+    braced, parenthesized, parse_macro_input,
+};
+use syn::parse::{Result, Parse, ParseStream};
+use syn::punctuated::Punctuated;
+use quote::quote;
+use crate::tt::TS;
+
+struct IdentOrWild(Ident);
+
+impl Parse for IdentOrWild {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        Ok(if input.peek(Token![_]) {
+            input.parse::<Token![_]>()?;
+            IdentOrWild(Ident::new("_", Span::call_site()))
+        } else {
+            IdentOrWild(input.parse()?)
+        })
+    }
+}
+
+enum QueryAttribute {
+    Desc(Option<Ident>, Punctuated<Expr, Token![,]>),
+    Cache(Option<Ident>, Expr),
+    LoadCached(Ident, Ident, Block),
+    FatalCycle,
+}
+
+impl Parse for QueryAttribute {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let attr: Ident = input.parse()?;
+        if attr == "desc" {
```

Comment:
Please put a comment inside each of these blocks showing an example input that would be parsed. It's hard to read off from the parsing code.